pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+	}
+
+	hasAnyError := false
+	for _, response := range res.GetStatuses() {
+		if entryID, ok := entryIDIndexMap[response.EntryID]; ok {
+			switch response.GetStatus() {
+			case runtimev1pb.TopicEventResponse_SUCCESS: //nolint:nosnakecase
+				// on uninitialized status, this is the case it defaults to as an uninitialized status defaults to 0 which is
+				// success from protobuf definition
+				bulkSubDiag.statusWiseDiag[string(pubsub.Success)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = nil
+			case runtimev1pb.TopicEventResponse_RETRY: //nolint:nosnakecase
+				bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = errors.Errorf("RETRY status returned from app while processing pub/sub event for entry id: %v", response.EntryID)
+				hasAnyError = true
+			case runtimev1pb.TopicEventResponse_DROP: //nolint:nosnakecase
+				log.Warnf("DROP status returned from app while processing pub/sub event for entry id: %v", response.EntryID)
+				bulkSubDiag.statusWiseDiag[string(pubsub.Drop)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = nil
+			default:
+				// Consider unknown status field as error and retry
+				bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = errors.Errorf("unknown status returned from app while processing pub/sub event  for entry id %v: %v", response.EntryID, response.GetStatus())
+				hasAnyError = true
+			}
+		} else {
+			log.Warnf("Invalid entry id received from app while processing pub/sub event %v", response.EntryID)
+			continue
+		}
+	}
+	for _, item := range msg.entries {
+		ind := entryIDIndexMap[item.EntryID]
+		if (*bulkResponses)[ind].EntryID == "" {
+			(*bulkResponses)[ind].EntryID = item.EntryID
+			(*bulkResponses)[ind].Error = errors.Errorf("Response not received, RETRY required while processing bulk subscribe event for entry id: %v", item.EntryID)
+			hasAnyError = true
+			bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+		}
+	}
+	bulkSubDiag.elapsed = elapsed
+	if hasAnyError {
+		return errors.New("Few message(s) have failed during bulk subscribe operation")
+	} else {
+		return nil
+	}
+}
+
+// sendBulkToDeadLetter sends the bulk message to deadletter topic.
+func (a *DaprRuntime) sendBulkToDeadLetter(
+	ctx context.Context, name string, msg *pubsub.BulkMessage, deadLetterTopic string,
+	entryIDIndexMap *map[string]int, bulkResponses *[]pubsub.BulkSubscribeResponseEntry,
+	bulkSubDiag *bulkSubIngressDiagnostics,
+) error {
+	data := make([]pubsub.BulkMessageEntry, len(msg.Entries))
+
+	if bulkResponses == nil {
+		data = msg.Entries
+	} else {
+		n := 0
+		for _, message := range msg.Entries {
+			entryID := (*entryIDIndexMap)[message.EntryID]
+			if (*bulkResponses)[entryID].Error != nil {
```

Comment:
Reviewers: If bulkResponses array is sent, it is checked what all messages errored, only sends to dlq the errored messages.